core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java

```diff
@@ -74,30 +69,33 @@ final class CompositeEndpointGroup extends AbstractEndpointGroup implements List
                 CompletableFuture.anyOf(this.endpointGroups.stream()
                                                            .map(EndpointGroup::whenReady)
                                                            .toArray(CompletableFuture[]::new))
-                                 .thenApply(unused -> endpoints());
+                                 .thenApply(unused -> rebuildEndpoints());
 
         this.selectionStrategy = requireNonNull(selectionStrategy, "selectionStrategy");
         selector = requireNonNull(selectionStrategy, "selectionStrategy").newSelector(this);
     }
 
     @Override
     public List<Endpoint> endpoints() {
-        if (!dirty.get()) {
-            return merged;
-        }
-
-        if (!dirty.compareAndSet(true, false)) {
-            // Another thread might be updating merged at this time, but endpoint groups are allowed to take a
-            // little bit of time to reflect updates.
-            return merged;
-        }
+        final List<Endpoint> endpoints = merged.get();
+        assert endpoints != null;
+        return endpoints;
+    }
 
-        final ImmutableList.Builder<Endpoint> newEndpoints = ImmutableList.builder();
-        for (EndpointGroup endpointGroup : endpointGroups) {
-            newEndpoints.addAll(endpointGroup.endpoints());
+    private List<Endpoint> rebuildEndpoints() {
+        for (;;) {
+            // Get the current endpoints before making a new endpoints list.
+            final List<Endpoint> oldEndpoints = merged.get();
+            final ImmutableList.Builder<Endpoint> newEndpointsBuilder = ImmutableList.builder();
+            for (EndpointGroup endpointGroup : endpointGroups) {
+                newEndpointsBuilder.addAll(endpointGroup.endpoints());
+            }
+            final List<Endpoint> newEndpoints = newEndpointsBuilder.build();
+            if (merged.compareAndSet(oldEndpoints, newEndpoints)) {
```

Comment:
Because `ImmutableList.of()` returns a singleton instance, `rebuildEndpoints` may update outdated endpoints in rare cases.
1. Thread A updates `endpointA`.
  - Thread A prepares to set empty -> `endpointA` in L93
2. Thread B removes `endpointA`
  - Thread B sets empty -> empty in L94
3. Thread A sets empty -> `endpointA` in L94
4. merged is `list(endpointA)` but the actual endpoints in `endpointGroups` are empty.